### PR TITLE
Only execute one final reduction in InternalAutoDateHistogram

### LINKED ISSUE #44914: auto_date_histogram fails where date_histogram does not
We are planning to port most of Stack Monitoring current aggregations to use `auto_date_histogram` in lieu of `date_histogram` as part of https://github.com/elastic/kibana/issues/37246.

We noticed that `auto_date_histogram` supports most of aggregations that `date_histogram` does, but fails (on `fetch` phase) with `bucket_script`. As a first pass, I tried replacing the existing `date_histogram` usage with `auto_date_histogram` and it broke:

```http
GET .monitoring-logstash-*/_search
{
  "size": 0,
  "query": {
    "bool": {
      "filter": [
        {
          "range": {
            "logstash_stats.timestamp": {
              "format": "epoch_millis",
              "gte": 1563551274281,
              "lte": 1563554874281
            }
          }
        }
      ]
    }
  },
  "aggs": {
    "check": {
      "auto_date_histogram": {
        "field": "logstash_stats.timestamp",
        "buckets": 10
      },
      "aggs": {
        "pipelines_nested": {
          "nested": {
            "path": "logstash_stats.pipelines"
          },
          "aggs": {
            "by_pipeline_id": {
              "terms": {
                "field": "logstash_stats.pipelines.id",
                "size": 1000
              },
              "aggs": {
                "by_pipeline_hash": {
                  "terms": {
                    "field": "logstash_stats.pipelines.hash",
                    "size": 1000
                  },
                  "aggs": {
                    "by_ephemeral_id": {
                      "terms": {
                        "field": "logstash_stats.pipelines.ephemeral_id",
                        "size": 1000
                      },
                      "aggs": {
                        "events_stats": {
                          "stats": {
                            "field": "logstash_stats.pipelines.events.out"
                          }
                        },
                        "throughput": {
                          "bucket_script": {
                            "script": "params.max - params.min",
                            "buckets_path": {
                              "min": "events_stats.min",
                              "max": "events_stats.max"
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

This query fails where the same query works if you replace `auto_date_histogram` with `date_histogram` (and use `interval` instead of `buckets`). Anyone attempting to run this on an existing cluster will likely need to shift the date `range` query.

### PR DESCRIPTION
**Pull request by polyfractal**
Because auto-date-histo can perform multiple reductions while merging buckets, we need to ensure that the intermediate reductions are done with a `finalReduce` set to false to prevent Pipeline aggs from generating their output.

This is done by creating a "fake" non-final reduce context for the intermediate merging.  Once all the buckets have been merged and the output is stable, a mostly-noop final reduction can be performed which will allow pipelines to generate their output.

Closes https://github.com/elastic/elasticsearch/issues/44914

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalAutoDateHistogram.java [hunk 1]
```diff
@@ -500,15 +500,24 @@ public InternalAggregation doReduce(List<InternalAggregation> aggregations, Redu
         BucketReduceResult reducedBucketsResult = reduceBuckets(aggregations, reduceContext);
 
         if (reduceContext.isFinalReduce()) {
+            // Because auto-date-histo can perform multiple reductions while merging buckets, we need to pretend this is
+            // not the final reduction to prevent pipeline aggs from creating their result early.  However we want
+            // to reuse the multiBucketConsumer so that max_buckets breaker is correctly accounted for
+            ReduceContext fakeReduceContext = new ReduceContext(reduceContext.bigArrays(), reduceContext.scriptService(),
```
Comment (pcsanwald) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalAutoDateHistogram.java]: I wonder if `penultimateReduceContext` or `stepBeforeFinalReduceContext` is clearer? on first glance, the term `fake` could read like it's a mistake something, I feel it implies something like "mocked reduce context" but that's not what's going on here. thoughts?

**Review by pcsanwald — APPROVED** [review]
one small comment but the rest of it LGTM

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalAutoDateHistogram.java (hunk 1 above)
Comment (polyfractal) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalAutoDateHistogram.java]: ++ you're right, that name could be confusing to future viewers.  Will give it something more descriptive (I rather like "penultimate"!)

**Comment by polyfractal** [comment]
@elasticmachine update branch

**Comment by polyfractal** [comment]
I reverted this commit because `testReduceRandom()` was consistently showing incorrect bucket counts compared to the max_buckets breaker in about 1/100 tests.

Unsure of the cause, but it was starting to look like it's not just a test bug but an actual problem with the PR.  Will revisit and open a new PR.

**Pull request merged**